doc/man3/EVP_PKEY_CTX_ctrl.pod

```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
 B<NID_ffdhe4096>, B<NID_ffdhe6144> or B<NID_ffdhe8192>.  This macro can be
 called during parameter or key generation.
 
+The EVP_PKEY_CTX_set_dh_rfc5114() and EVP_PKEY_CTX_set_dhx_rfc5114() set the
+RFC5114 parameters instead of generating new DH parameters. Possible values are
+1, 2 and 3 corresponding to RFC5114 sections 2.1, 2.2 and 2.3. If set all other
+DH parameter options are overriden. 0 can be used to clear the stored value.
+
+The EVP_PKEY_CTX_set_dh_kdf_type() macro sets the key derivation function type
+to B<kdf> for DH key derivation. Possible values are B<EVP_PKEY_DH_KDF_NONE>
+and B<EVP_PKEY_DH_KDF_X9_42>.
+
+The EVP_PKEY_CTX_set0_dh_kdf_oid() macro sets the key derivation function
+object identifier to B<oid> for DH key derivation.
+
+The EVP_PKEY_CTX_set_dh_kdf_md() macro sets the key derivation function
+message digest to B<md> for DH key derivation.
+
+The EVP_PKEY_CTX_set_dh_kdf_outlen() macro sets the key derivation function
+output length to B<len> for DH key derivation.
+
+The EVP_PKEY_CTX_set0_dh_kdf_ukm() macro sets the user key material to
+B<ukm> for DH key derivation.
+
 The EVP_PKEY_CTX_set_ec_paramgen_curve_nid() sets the EC curve for EC parameter
 generation to B<nid>. For EC parameter generation this macro must be called
 or an error occurs because there is no default curve.
 This function can also be called to set the curve explicitly when
 generating an EC key.
 
-The EVP_PKEY_CTX_set_ec_param_enc() sets the EC parameter encoding to
+The EVP_PKEY_CTX_set_ec_param_enc() macro sets the EC parameter encoding to
 B<param_enc> when generating EC parameters or an EC key. The encoding can be
 B<OPENSSL_EC_EXPLICIT_CURVE> for explicit parameters (the default in versions
 of OpenSSL before 1.1.0) or B<OPENSSL_EC_NAMED_CURVE> to use named curve form.
 For maximum compatibility the named curve form should be used. Note: the
 B<OPENSSL_EC_NAMED_CURVE> value was only added to OpenSSL 1.1.0; previous
 versions should use 0 instead.
 
+The EVP_PKEY_CTX_set_ecdh_cofactor_mode() macro sets the cofactor mode to
+B<cofactor_mode> for ECDH key derivation. Possible values are 1 to enable
+cofactor generation, 0 to disable it and -1 to clear the stored cofactor key.
+
+The EVP_PKEY_CTX_set_ecdh_kdf_type() macro sets the key derivation function type
+to B<kdf> for ECDH key derivation. Possible values are B<EVP_PKEY_ECDH_KDF_NONE>
+and B<EVP_PKEY_ECDH_KDF_X9_62>.
```

Comment:
This KDF seems particularly badly named! AFAICT this is actually the KDF from X9.63 not X9.62!!! X9.62 is ECDSA but X9.63 is ECDH (the same KDF is also specified in SEC1). If this option is used then the kdf_outlen and md must also be set.

Reply:
Perhaps we should create a synonym for this option called `EVP_PKEY_ECDH_KDF_X9_63` and use that in preference in the docs and code.

Reply:
Again this appears to be more than just a typo, the implementation in ecdh_kdf is also named ECDH_KDF_X9_62. Are we sure this is an error?

Reply:
Yes, I can see it is consistently referred to as the X9.62 KDF throughout the implementation, the API and the original commit message (25af7a5db). However I've looked through X9.62 and I can't see a KDF in it. Comparing the implementation against the X9.63 KDF spec (section 5.6.3) it does seem to match. To be on the safe side I've reached out to the original author (no longer working on the project) to see what he says. I'll let you know.

Reply:
The answer I got back on this was:

>The derivation is part of CMS support for ECDH which is defined in RFC5753 et
al. There should be some OIDs in there too which hopefully have X9.63 in them.
It was part of an effort to get DH support for CMS included: I think the only
thing that supported it at the time was Bouncy Castle.
>
>So yes it looks like there is an off by .01 error there.
>
>I can't double check at the moment: I'm away from my main PC and will be for
some days.